### PR TITLE
fix: Add ACL checks to Struts endpoints

### PR DESCRIPTION
**Pull request by netroms**
Adds acl checks on Struts endpoints.

[[DHIS2-10709](https://jira.dhis2.org/browse/DHIS2-10709)]

Signed-off-by: Morten Svanaes <msvanaes@dhis2.org>

### CODE REVIEW

File: dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/commons/action/BaseAction.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.hibernate.exception.ReadAccessDeniedException;
+import org.hisp.dhis.security.acl.AclService;
+import org.hisp.dhis.user.CurrentUserService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * @author Morten Svanæs <msvanaes@dhis2.org>
+ */
+public class BaseAction
```
Comment (jbee) [inline dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/commons/action/BaseAction.java:39]: should be `abstract` I think to make clear it is not used "as is".

File: dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/commons/action/BaseAction.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+{
+    @Autowired
+    protected AclService aclService;
+
+    @Autowired
+    protected CurrentUserService currentUserService;
+
+    public <T extends IdentifiableObject> void canReadType( Class<T> type )
```
Comment (jbee) [inline dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/commons/action/BaseAction.java:47]: I like to make such helpers `final` to reduce number of methods that come up in IDE suggestions and to make sure this logic isn't bypassed.

**Review by jbee — APPROVED** [review]
I was scared by the number of files changed but this wasn't so bad :D

Just some minor code style suggestions.

File: dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/commons/action/BaseAction.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        if ( !aclService.canRead( currentUserService.getCurrentUser(), type ) )
+        {
+            throw new ReadAccessDeniedException(
+                "You don't have the proper permissions to read objects of this type." );
+        }
+    }
+
+    public void canReadInstance( IdentifiableObject instance )
+    {
+        if ( !aclService.canRead( currentUserService.getCurrentUser(), instance ) )
```
Comment (vietnguyen) [inline dhis-2/dhis-web/dhis-web-commons/src/main/java/org/hisp/dhis/commons/action/BaseAction.java:58]: `currentUserService.getCurrentUser()` might be called multiple time if this `canRead*()` is called in a loop.
Maybe consider to add another method `canRead*( User user, IdentifiableObject instance)` and move `getCurrentUser()` to upper layer.

**Review by vietnguyen — APPROVED** [review]

**Comment by larshelge** [comment]
This looks good. I assume, this is for sending an explicit access denied message response? The metadata read check is already "built into" the service layer, which will return null both if the object does not exist and if the user does not have read access. So we want to differentiate and then send access denied in the response instead not found?

**Review by jbee — APPROVED** [review]

**Pull request merged**